### PR TITLE
Replace CWallet::Set* functions that use memonly with Add/Load variants

### PR DESCRIPTION
**Pull request by achow101**
`SetHDChaiin`, `SetActiveScriptPubKeyMan`, and `SetWalletFlags` have a `memonly` argument which is kind of confusing, as noted in https://github.com/bitcoin/bitcoin/pull/17681#discussion_r427633081. This PR replaces those functions with `Add*` and `Load*` variants so that they follow the pattern used elsewhere in the wallet.

`AddHDChain`, `AddActiveScriptPubKeyMan`, and `AddWalletFlags` both set their respective variables in `CWallet` and writes them to disk. These functions are used by the actions which modify the wallet such as `sethdseed`, `importdescriptors`, and creating a new wallet.

`LoadHDChain`, `LoadActiveScriptPubKeyMan`, and `LoadWalletFlags` just set the `CWallet` variables. These functions are used by `LoadWallet` when loading the wallet from disk.

### CODE REVIEW

**Comment by laanwj** [comment]
I agree that having separate Add/Load functions instead of a confusing boolean argument improves understandability of the code quite a bit.
Code review ACK f123374a661d54e091108e490f8c4cc9ebc0d251

**Comment by Empact** [comment]
Concept ACK

File: src/wallet/wallet.cpp [hunk 1]
```diff
@@ -1385,19 +1385,26 @@ bool CWallet::IsWalletFlagSet(uint64_t flag) const
...
     }
-    if (!memonly && !WalletBatch(*database).WriteWalletFlags(m_wallet_flags)) {
+
+    return true;
+}
+
+bool CWallet::AddWalletFlags(uint64_t flags)
+{
+    LOCK(cs_wallet);
+    if (!WalletBatch(*database).WriteWalletFlags(flags)) {
```
Comment (Empact) [inline src/wallet/wallet.cpp:1403]: In the existing code, on Add, the `contains unknown non-tolerable wallet flags` test and bail will occur first and prevent writing the database. With this change, the values are written, and the failure is only identified after, which seems improper.

Comment (Empact) [reply src/wallet/wallet.cpp:1403]: Perhaps add a test for this case?

Comment (achow101) [reply src/wallet/wallet.cpp:1403]: That failure is only relevant when loading a wallet from disk. `AddWalletFlags` is only used internally in wallet creation; we would never add flags that we don't know.

**Review by ryanofsky — APPROVED** [review]
Code review ACK f123374a661d54e091108e490f8c4cc9ebc0d251. Contra https://github.com/bitcoin/bitcoin/pull/19046#discussion_r430710261 it seems good to me that these methods now try to update the wallet database first and return early on failure before updating the state in memory. Haven't thought about it too much but it seems better if that there's an error writing to disk for some reason, operation just fails and wallet stays in the current state.

@jnewbery, you may be interested in this PR since it was inspired by your review comment

File: src/wallet/wallet.cpp (hunk 1 above)
Comment (jnewbery) [reply src/wallet/wallet.cpp:1403]: I agree that this seems less safe than the existing code. Is it possible to maintain the current order of checking first and then writing to disk?

Comment (ryanofsky) [reply src/wallet/wallet.cpp:1403]: > I agree that this seems less safe than the existing code.

In current code m_wallet_flags can be updated even if database write fails leading to inconsistency between disk and memory state and possibility for data inconsistent with flags to subseqently be written to wallet. New code writing to disk first, memory second, prevents this type of issue and would seem to be more safe. What's the scenario where it's less safe?

Comment (jnewbery) [reply src/wallet/wallet.cpp:1403]: If this function is called with an unknown wallet flag, it'll be written to disk and then won't be applied to the in-memory wallet (and won't be able to be loaded later), which is also a data inconsistency.

The unknown flag check was added to protect when loading a wallet from a newer version on an older version, so I think your concern is more valid. I don't see any harm in a sanity assert that we don't try to add an unknown flag to the wallet before writing to disk.

Comment (achow101) [reply src/wallet/wallet.cpp:1405]: I've added an assert in `AddWalletFlags` to handle that case.

File: src/wallet/wallet.cpp [hunk 2]
```diff
@@ -1385,19 +1385,28 @@ bool CWallet::IsWalletFlagSet(uint64_t flag) const
...
-    if (!memonly && !WalletBatch(*database).WriteWalletFlags(m_wallet_flags)) {
+    m_wallet_flags = flags;
+
+    return true;
+}
+
+bool CWallet::AddWalletFlags(uint64_t flags)
+{
+    LOCK(cs_wallet);
+    // We should never be writing unknown onon-tolerable wallet flags
```
Comment (jnewbery) [inline src/wallet/wallet.cpp:1403]: s/onon/non/

**Review by jnewbery — COMMENTED** [review]
Code review ACK 3a9aba21a49a6d80bd187940d5e26893937b6832

One small typo in a comment if you feel like retouching.

File: src/wallet/wallet.cpp [hunk 3]
```diff
@@ -1385,19 +1385,28 @@ bool CWallet::IsWalletFlagSet(uint64_t flag) const
...
+    assert(!(((flags & KNOWN_WALLET_FLAGS) >> 32) ^ (flags >> 32)));
```
Comment (ryanofsky) [inline src/wallet/wallet.cpp:1404]: In commit "Split SetWalletFlags into Add/LoadWalletFlags" (3a9aba21a49a6d80bd187940d5e26893937b6832)

If you wind up updating check may be more readable as `assert(((flags & KNOWN_WALLET_FLAGS) >> 32) == (flags >> 32))`

**Review by ryanofsky — APPROVED** [review]
Code review ACK 3a9aba21a49a6d80bd187940d5e26893937b6832. Only changes since last review tweaks making m_wallet_flags updates more safe

File: src/wallet/wallet.cpp (hunk 2 above)
Comment (achow101) [reply src/wallet/wallet.cpp:1403]: Oops. I'll fix that if I have to push again.

**Review by meshcollider — COMMENTED** [review]
utACK 3a9aba21a49a6d80bd187940d5e26893937b6832

**Pull request merged**